keps/sig-node/2238-liveness-probe-grace-period/README.md

```diff
@@ -0,0 +1,577 @@
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+
+* **What steps should be taken if SLOs are not being met to determine the problem?**
+
+[supported limits]: https://git.k8s.io/community//sig-scalability/configs-and-limits/thresholds.md
+[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
+
+## Implementation History
+
+- 2021-01-07: Initial draft KEP
+
+## Drawbacks
+
+The main drawback of this proposal is that it requires a (compatible) API
+change.
+
+## Alternatives
+
+From [#64715](https://github.com/kubernetes/kubernetes/issues/64715#issuecomment-756173979):
+
+Infer the maximum grace period for liveness from `num failures * probe period`,
+implying that a user with short liveness reaction wants a fast reaction from
+their pod (@smarterclayton)
+
+- PRO: Any user unaware of the coupling reacts to liveness failures faster,
+  normal pods (including reason 2 above) are less likely to be impacted
+- CON: If we later want to change the API to indicate this field, the default
+  behavior makes less sense (i.e. complicates choosing the option above)
+- CON: A user with a long grace period who wants liveness to take more than
+  tens of seconds (not sure why they would) suddenly sees that behavior
+  impacted
+- CON: This sort of implicit calculation feels pretty arbitrary (we generally
+  make defaults explicit vs calculated)
+
+Introduce a new toleration, similar to execute toleration for node not ready
+(how long the pod should be left on the node while unready before being
+evicted), that controls how long liveness tolerates (@smarterclayton)
+
+- PRO: avoids API change
+- CON: basically a hack to avoid an API change, tolerations for execution don't
+  really make sense for conditional stuff
+
+Since liveness probe failures indicate that a container is unhealthy and should
+be terminated, update the behaviour of liveness probes to terminate the pod
+immediately, or with some default grace period (@sjenning)
+
+- PRO: avoids API change
+- CON: may break existing applications that rely on the grace period for crash
+  safety
+- CON: thresholds not user-configurable
+- CON: even if we use a feature flag for this, it will eventually become the
+  required behaviour
+
+Introduce a similar field, but at the pod-level, next to the existing field.
+
+- PRO: similar values are grouped together
+- CON: logically, liveness probes operate on a per-container basis, and thus we
```

Comment:
so the grace period will sum up grace periods of all containers? I didn't get this CON rationale =(

Reply:
The pod-level termination grace period makes sense to me because that's the overall time it should take to shut down the whole pod as a group of containers. But liveness probes don't act at the pod level, they act exclusively at the container level. So it doesn't make as much sense to me to set a global liveness timeout: what if you wanted to set them to different values for different containers?

Reply:
so this is the time to restart the specific container in pod? What if the policy is restart never and pod needs to be terminated. Which timeout will be used?

Reply:
Yes, liveness probes are specific to each container in a pod.

If the policy is restart never and the whole pod needs to be terminated, the other containers would either use the default grace period or, if set, the pod-level terminationGracePeriodSeconds.

Reply:
that explains it. I misunderstood the KEP =) Will read again with this understanding.

Immediate question though, would users expect that liveness grace periods will be respected for each pod when overall pod needs to be terminated?

Reply:
Nope. The idea of this KEP is to differentiate between the grace period for a regular, graceful shutdown when pod termination is requested (happy path), and a liveness probe failure (sad path), because these have been accidentally coupled.